File: codes/Garnetwzy/295.cpp
class MedianFinder {
public:
    MedianFinder() {
        
    }
    
    void addNum(int num) {
        if (left.empty() || num <= left.top()) {
            left.push(num);
        } else {
            right.push(num);
        }
        
        if (left.size() > right.size() + 1) {
            right.push(left.top());
            left.pop();
        }
        
        if (right.size() > left.size() + 1) {
            left.push(right.top());
            right.pop();
        }
    }
    
    double findMedian() {
        if (left.size() > right.size()) 
            return left.top();
        if(left.size() < right.size())
            return right.top();
        return (left.top() + right.top()) / 2.0;
    }
    
private:
        /** initialize your data structure here. */
    priority_queue<int> left;
    priority_queue<int, vector<int>, greater<int>> right;
};

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder* obj = new MedianFinder();
 * obj->addNum(num);
 * double param_2 = obj->findMedian();
 */

File: codes/HellDream/295.java
class MedianFinder {

    /** initialize your data structure here. */
    PriorityQueue<Integer> max  = new PriorityQueue<>();
    PriorityQueue<Integer> min = new PriorityQueue<>(Collections.reverseOrder());
    boolean even = true;
    public MedianFinder() {
        
    }
    
    public void addNum(int num) {
        if(even){
            max.offer(num);
            min.offer(max.poll());
        }else {
            min.offer(num);
            max.offer(min.poll());
        }
        even = !even;
    }
    
    public double findMedian() {
        if(even){
            return (max.peek()+min.peek())/2.0;
        }
        return min.peek();
    }
}